File: public/dsa/searching/contains_element/optimal/contains_element.cpp
#include <bits/stdc++.h>
using namespace std;

bool containsElement(vector<int>& arr, int key) {
    int left = 0, right = arr.size() - 1;
    while (left <= right) {
        int mid = (left + right) / 2;
        if (arr[mid] == key) {
            return true;
        } else if (arr[mid] < key) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return false;
}

// tests-start
int main() {
    vector<int> arr = {1, 2, 3, 3, 3, 4, 4, 5};
    int key = 2;

    cout << (containsElement(arr, key) ? "true" : "false") << endl;

    return 0;
}
// tests-end

/*output
true
*/

File: public/dsa/searching/contains_element/optimal/contains_element.py
from typing import List


class Solution:
    def containsElement(self, arr: List[int], key: int) -> bool:
        left, right = 0, len(arr) - 1
        while left <= right:
            mid = (left + right) // 2
            if arr[mid] == key:
                return True
            elif arr[mid] < key:
                left = mid + 1
            else:
                right = mid - 1
        return False


# tests-start
arr = [1, 2, 3, 3, 3, 4, 4, 5]
key = 2
sol = Solution()
sol.containsElement(arr, key)
# tests-end

'''output
True
'''
